### PR TITLE
code clean up in the glusterd2 plugins

### PR DESCRIPTION
**Pull request by Madhu-1**
* Added logging with the log.WithError()
* allocate capacity to slice to reduce the allocation of memory during append
*added UNDO function in txn steps

signed-off-by: Madhu Rajanna <mrajanna@redhat.com>

### CODE REVIEW

**Comment by Madhu-1** [comment]
@prashanthpai  @aravindavk please review

**Comment by Madhu-1** [comment]
retest this place

**Comment by Madhu-1** [comment]
retest this please

File: plugins/georeplication/rest.go [hunk 1]
```diff
@@ -171,11 +174,18 @@ func georepCreateHandler(w http.ResponseWriter, r *http.Request) {
 	// Workaround till {{ volume.id }} added to the marker options table
 	vol.Options["marker.volume-uuid"] = vol.ID.String()
 
+	//save volume information for transaction failure scenario
+	if err := txn.Ctx.Set("oldvolinfo", oldvolinfo); err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
```
Comment (rishubhjain) [inline plugins/georeplication/rest.go]: @Madhu-1 No logging required?

Comment (Madhu-1) [reply plugins/georeplication/rest.go]: added logging, In most of the cases we are not logging if txn.Set fails. I will improve logging in upcoming PR's

File: plugins/georeplication/rest.go [hunk 2]
```diff
@@ -171,11 +174,19 @@ func georepCreateHandler(w http.ResponseWriter, r *http.Request) {
 	// Workaround till {{ volume.id }} added to the marker options table
 	vol.Options["marker.volume-uuid"] = vol.ID.String()
 
+	//save volume information for transaction failure scenario
+	if err := txn.Ctx.Set("oldvolinfo", oldvolinfo); err != nil {
+		logger.WithError(err).Error("failed to set oldvolinfo in transaction context")
```
Comment (rishubhjain) [inline plugins/georeplication/rest.go]: Add WithField("key", "oldvolinfo")

Comment (rishubhjain) [reply plugins/georeplication/rest.go]: TRIVIAL: Also try to use mixedCases, oldvolinfo -> oldVolInfo

Comment (Madhu-1) [reply plugins/georeplication/rest.go]: I think we are not currently not following the convention of using mixedCases in glusterd2, and we should add this rule to the development guide.

File: plugins/georeplication/gsyncd.go [hunk 3]
```diff
@@ -37,7 +37,7 @@ func (g *Gsyncd) Path() string {
 
 // Args returns arguments to be passed to gsyncd process during spawn.
 func (g *Gsyncd) Args() []string {
-	g.args = []string{}
+	g.args = make([]string, 0, 6)
```
Comment (prashanthpai) [inline plugins/georeplication/gsyncd.go]: Use a specific size only if it won't change in future or if length is computed using `len()`. The args here may change in future, might become 5 or 7 which requires this to be changed. So, please don't do this

`premature optimization is the root of all evil`

File: plugins/georeplication/rest.go [hunk 4]
```diff
@@ -1110,8 +1097,7 @@ func georepSSHKeyPushHandler(w http.ResponseWriter, r *http.Request) {
 
 	err = txn.Do()
 	if err != nil {
-		logger.WithFields(log.Fields{
-			"error":   err.Error(),
+		logger.WithError(err).WithFields(log.Fields{
```
Comment (prashanthpai) [inline plugins/georeplication/rest.go]: Use `WithField` if it's only one field. Here and everywhere else.

File: plugins/georeplication/transactions.go [hunk 5]
```diff
@@ -255,7 +255,7 @@ func configFileGenerate(session *georepapi.GeorepSession) error {
 	}
 
 	// Remote host and UUID details
-	var remote []string
+	var remote = make([]string, 0, len(session.RemoteHosts))
```
Comment (prashanthpai) [inline plugins/georeplication/transactions.go]: Again, premature optimization. Seems like you are too much focused on cleanup even when it's not necessarry

File: plugins/georeplication/transactions.go [hunk 6]
```diff
@@ -264,8 +264,9 @@ func configFileGenerate(session *georepapi.GeorepSession) error {
 	)
 
 	// Master Bricks details
-	var master []string
-	for _, b := range vol.GetBricks() {
+	bricks := vol.GetBricks()
```
Comment (prashanthpai) [inline plugins/georeplication/transactions.go]: Same here.

File: plugins/georeplication/store-utils.go [hunk 7]
```diff
@@ -42,7 +42,7 @@ func getSession(masterid string, remoteid string) (*georepapi.GeorepSession, err
 func addOrUpdateSession(v *georepapi.GeorepSession) error {
 	json, e := json.Marshal(v)
 	if e != nil {
-		log.WithField("error", e).Error("Failed to marshal the Info object")
+		log.WithError(e).Error("Failed to marshal the Info object")
```
Comment (prashanthpai) [inline plugins/georeplication/store-utils.go]: You have two commits which does error related fixed. Merge all of them into single commit

File: plugins/glustershd/glustershd.go [hunk 8]
```diff
@@ -47,7 +47,7 @@ func (shd *Glustershd) Args() []string {
 	glusterdSockDir := config.GetString("rundir")
 	socketfilepath := fmt.Sprintf("%s/%x.socket", glusterdSockDir, xxhash.Sum64String(gdctx.MyUUID.String()))
 
-	shd.args = []string{}
+	shd.args = make([]string, 0, 6)
```
Comment (prashanthpai) [inline plugins/glustershd/glustershd.go]: See other comment

File: plugins/georeplication/transactions.go (hunk 5 above)
Comment (Madhu-1) [reply plugins/georeplication/transactions.go]: I think this is necessary to have it. leaving as it is

File: plugins/georeplication/transactions.go (hunk 6 above)
Comment (Madhu-1) [reply plugins/georeplication/transactions.go]: same here

File: plugins/georeplication/store-utils.go (hunk 7 above)
Comment (Madhu-1) [reply plugins/georeplication/store-utils.go]: done

File: plugins/glustershd/glustershd.go (hunk 8 above)
Comment (Madhu-1) [reply plugins/glustershd/glustershd.go]: done ,removed hard coding

File: plugins/georeplication/gsyncd.go (hunk 3 above)
Comment (Madhu-1) [reply plugins/georeplication/gsyncd.go]: done

File: plugins/georeplication/rest.go (hunk 4 above)
Comment (Madhu-1) [reply plugins/georeplication/rest.go]: am using ```WithFields``` if I have multiple fields to log.
am not getting where I have used ```WithFields``` if I have one field to log.

**Pull request merged**